### DISCUSSION TITLE
Two more issues in running custom+pcp_combine+grid_stat

### DISCUSSION THREAD

**Discussion by lesnow93** [dtcenter/METplus#1891] [Configuration] [answered]
Hello again!  I am continuing to make progress in my test case running pcp_combine and grid_stat in a multi-model, custom configuration.  I am now successful in processing the multi-model set up, but there are 2 remaining problems occurring:

1) I specify BUDGET for interpolating APCP to the OBS grid to occur within grid_stat (so as to not have to run regrid_data_plaane), but the MET OUTPUT files are defaulting to NEAREST, appearing to ignore the settings I have in my config file:

GRID_STAT_REGRID_TO_GRID = OBS
GRID_STAT_REGRID_METHOD = BUDGET
GRID_STAT_REGRID_WIDTH = 2
GRID_STAT_REGRID_SHAPE = SQUARE
GRID_STAT_REGRID_VLD_THRESH = 0.5

2) I have my config file set up to loop through A06, A12, and A24 accumulated precip intervals.  It successfully runs through pcp_combine for all the intervals; however, the job errors out after processing the A06 in grid_stat. 

OBS_VAR1_NAME = APCP
OBS_VAR1_LEVELS = A06, A12, A24
FCST_VAR1_NAME = APCP
FCST_VAR1_LEVELS = A06, A12, A24

I attached the latest config file and a log file for only 2 of the custom loop members (for expediency in testing)
[hiwat_multiModel_GridStat_APCP_v3.conf.txt](https://github.com/dtcenter/METplus/files/9861669/hiwat_multiModel_GridStat_APCP_v3.conf.txt)
[metplus_201803291800_v3.log](https://github.com/dtcenter/METplus/files/9861671/metplus_201803291800_v3.log)
.

Sincerely,
Jonathan

**Comment by georgemccabe** [comment]
Hi @lesnow93, I see the following message in your log file:

> DEBUG: Skip writing output /raid1/sport/people/casejl/METplus/hiwat_output/gridstat/2018032918/grid_stat_HKH1_APCP_vs_imergf_APCP_A06\*20180330_000000V\* because it already exists. Remove file or change GRID_STAT_SKIP_IF_OUTPUT_EXISTS to False to process

This means that you have the wrappers configured to skip running grid_stat because the output file already exists. This would explain why you aren't seeing the updated regrid method used in the output. Setting the following in your config file should resolve this issue:

`GRID_STAT_SKIP_IF_OUTPUT_EXISTS = False`

It looks like the same is true for PCPCombine, so you may also need to set:


`PCP_COMBINE_SKIP_IF_OUTPUT_EXISTS = False`

Also, it looks like your log file contains many runs of the METplus wrappers, so it is difficult to see what was run. Could you generate a fresh log file with the above changes and send it to me if you are still having issues?

**Reply by lesnow93** [reply]
I thought I uploaded a fresh log file, so I apologize if it had extra junk in there.  I did do a clean METplus run where I deleted all the pcp_combine and grid_stat output, but I still get NEAREST instead of BUDGET for the interpolation in the grid_stat output.   In fact, I never have gotten the BUDGET interpolation in the MET .stat and other output files. I'll try to upload a cleaner log file Thursday.

**Comment by georgemccabe** [comment]
Hi @lesnow93, I see what is happening now. The NEAREST setting is coming from smoothing, which is controlled by the interp dictionary. Here is the default value:
https://github.com/dtcenter/MET/blob/main_v10.1/met/data/config/GridStatConfig_default#L153
```
interp = {
   field      = BOTH;
   vld_thresh = 1.0;
   shape      = SQUARE;


   type = [
      {
         method = NEAREST;
         width  = 1;
      }
   ];
}
```

You can override this value by setting [GRID_STAT_INTERP_TYPE_METHOD](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-GRID_STAT_INTERP_TYPE_METHOD) = BUDGET.

**Reply by lesnow93** [reply]
Thanks for that answer, George.  Unfortunately, I'm still receiving 2 more errors in the output.  

The first problem is related to pcp_combine in handling subtraction of hours in the forecast data, which doesn't cause failure, but still reports errors.  For shorter duration forecast hours, it tries to use negative forecast hours when constructing the forecast filename, which seems like a bug in METplus.  

The second problem arose when I added GRID_STAT_INTERP_TYPE_METHOD = BUDGET to my config file.  I'm now receiving errors in grid_stat as follows:
ERROR  :
ERROR  : Dictionary::lookup_num_array() -> lookup failed for name "width"
ERROR  :

Latest config and log files attached.  
Thanks for the continued help,
Jonathan

[hiwat_multiModel_GridStat_APCP_v4.conf.txt](https://github.com/dtcenter/METplus/files/9882742/hiwat_multiModel_GridStat_APCP_v4.conf.txt)
[metplus_201803291800_v4.log](https://github.com/dtcenter/METplus/files/9882743/metplus_201803291800_v4.log)

**Reply by georgemccabe** [reply]
Hi Jonathan,

The 2nd issue sounds like you need to also set the interp.type.width if you set method, so you'll have to set GRID_STAT_INTERP_TYPE_WIDTH in your METplus config.

The first issue looks like it is trying to use a negative number for the 2nd accumulation for the subtract command. The 2nd accumulation is calculated by subtracting the desired accumulation from the forecast lead. From the input files I have seen, a 6 hour forecast lead file does not contain accumulations greater than 6, so it doesn't seem possible to compute at 12 hour accumulation with that file. Do your files contain accumulations greater than the forecast lead? If not, I don't think the subtract method of PCPCombine is what you want to use. I think you will want to use the add method to build the accumulation you need by adding accumulations together to build the larger accumulation. Does that make sense in this case? If not, you could send us some of your sample files so I can see how this would be configured.

**Reply by lesnow93** [reply]
Hello again George.

Regarding the negative forecast hours using the SUBTRACT method: My forecast output precip accumulations are stored as total model run accumulations (e.g., 20-h forecast file has 20-h total precip accumulation, 37-h has 37-h total precip, etc.).  So I have to use the subtract method to come up with accumulation intervals at any given forecast hour.  The negative forecast hours in the SUBTRACT method arise from the fact that I also want to compute stats for A06, A12, and A24, but I'm specifying "LEAD_SEQ = begin_end_incr(6, 48, 6)" in order to catch the first 6-h forecast time.  My follow-up question is therefore, how can I specify a LEAD_SEQ that will cover all the forecast accumulation intervals over which I want to verify?  Like I said above, the processes don't error out when pcp_combine encounters the negative forecast hours, so perhaps I can ignore it and move on.  But I'd like to have a clean error-less METplus run, after all!

-Jonathan

**Reply by georgemccabe** [reply]
Hi @lesnow93, thank you for that info. Do I understand correctly that you only need to compute the accumulations that are less than or equal the current forecast lead? For example, for a 6 hour lead, you want to compute the 6 hour accumulation; for a 12 hour lead, you want to compute 6 and 12 hour accumulations, etc.

If that is the case, then we could enhance the PCPCombine subtract method logic to warn and skip if a negative accumulation is encountered instead of attempting to substitute the negative value and error when the file is not found. This seems like a reasonable enhancement and would not take very long to implement. I created issue #1905 to track this work.

There is a config option called `INPUT_MUST_EXIST` that can be set to False to prevent errors if the input files are not found. This is primarily used internally for our unit tests and is not fully supported in every wrapper. This option is rarely used otherwise because it may also prevent errors that should not occur from being reported, therefore hiding bigger issues.

**Reply by lesnow93** [reply]
That is exactly right, George!  At any given forecast time, if an accumulation interval length exceeds the forecast hour, then the pcp_combine subtraction should not take place and thus, processing should not occur.  That is the logic I had in my crusty old version of a python wrapper for computing APCP with pcp_combine/subtract.

**Reply by georgemccabe** [reply]
OK, sounds like a good change to incorporate into the PCPCombine wrapper. Do you have any other on-going issues from this thread?

**Reply by lesnow93** [reply]
I commented on your new thread #1905 to clarify that the metplus logic is generating negative forecast hours when constructing the filename, not a negative accumulation interval.  

Other than that, I think this thread is done.  Regarding the regridding within grid_stat, I was getting confused with the INTERP method column in the .stat files (thinking that was the REGRID method).  So I'm convinced that the regridding is going well within grid_stat.  It's just that there isn't any output in the .stat columns to indicate that the regridding went as planned.

Best,
Jonathan

**Reply by georgemccabe** [reply]
Thank you for commenting on the new issue. Feel free to add any other important details you think are appropriate there. I created the issue to keep it on our radar but did not yet fill in all of the relevant context. The example you provided with the negative filename was exactly what I was planning on adding later, so I appreciate you doing that!

I'm going to lock this discussion to prevent future posts. You can track the development progress through the issue and feel free to create a new discussion if you have any other issues or questions.

**Comment by georgemccabe** [comment, marked as answer]
To summarize, issue #1905 was created to handle when the desired accumulation interval is less than or equal to the forecast lead in PCPCombine subtract method more appropriately.
